precompiles/assets-erc20/src/eip2612.rs

```diff
@@ -0,0 +1,275 @@
+macro_rules! impl_prefix {
+	($instance:ident, $name:literal) => {
+		// Using `paste!` we generate a dedicated module to avoid collisions
+		// between each instance `Nonces` struct.
+		paste::paste! {
+			mod [<_impl_prefix_ $instance:snake>] {
+				use super::*;
+
+				pub struct Nonces;
+
+				impl StorageInstance for Nonces {
+					const STORAGE_PREFIX: &'static str = "Nonces";
+
+					fn pallet_prefix() -> &'static str {
+						$name
+					}
+				}
+
+				impl InstanceToPrefix for $instance {
+					type NoncesPrefix = Nonces;
+				}
+			}
+		}
+	};
+}
+
+// Since the macro expect a `ident` to be used with `paste!` we cannot provide `()` directly.
+type Instance0 = ();
+
+impl_prefix!(Instance0, "Erc20Instance0Assets");
+impl_prefix!(Instance1, "Erc20Instance1Assets");
+impl_prefix!(Instance2, "Erc20Instance2Assets");
+impl_prefix!(Instance3, "Erc20Instance3Assets");
+impl_prefix!(Instance4, "Erc20Instance4Assets");
+impl_prefix!(Instance5, "Erc20Instance5Assets");
+impl_prefix!(Instance6, "Erc20Instance6Assets");
+impl_prefix!(Instance7, "Erc20Instance7Assets");
+impl_prefix!(Instance8, "Erc20Instance8Assets");
+impl_prefix!(Instance9, "Erc20Instance9Assets");
+impl_prefix!(Instance10, "Erc20Instance10Assets");
+impl_prefix!(Instance11, "Erc20Instance11Assets");
+impl_prefix!(Instance12, "Erc20Instance12Assets");
+impl_prefix!(Instance13, "Erc20Instance13Assets");
+impl_prefix!(Instance14, "Erc20Instance14Assets");
+impl_prefix!(Instance15, "Erc20Instance15Assets");
+impl_prefix!(Instance16, "Erc20Instance16Assets");
+
+/// Storage type used to store EIP2612 nonces.
+pub type NoncesStorage<Instance> = StorageDoubleMap<
+	<Instance as InstanceToPrefix>::NoncesPrefix,
+	Blake2_128Concat,
+	// Asset contract address
+	H160,
+	Blake2_128Concat,
+	// Owner
+	H160,
+	// Nonce
+	U256,
+	ValueQuery,
+>;
+
+pub struct Eip2612<Runtime, IsLocal, Instance: 'static = ()>(
+	PhantomData<(Runtime, IsLocal, Instance)>,
+);
+
+impl<Runtime, IsLocal, Instance> Eip2612<Runtime, IsLocal, Instance>
+where
+	Instance: InstanceToPrefix + 'static,
+	Runtime: pallet_assets::Config<Instance>
+		+ pallet_evm::Config
+		+ frame_system::Config
+		+ pallet_timestamp::Config,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::Call: From<pallet_assets::Call<Runtime, Instance>>,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	BalanceOf<Runtime, Instance>: TryFrom<U256> + Into<U256> + EvmData,
+	Runtime: AccountIdAssetIdConversion<Runtime::AccountId, AssetIdOf<Runtime, Instance>>,
+	<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin: OriginTrait,
+	IsLocal: Get<bool>,
+	<Runtime as pallet_timestamp::Config>::Moment: Into<U256>,
+{
+	fn compute_domain_separator(address: H160, asset_id: AssetIdOf<Runtime, Instance>) -> [u8; 32] {
+		let name: H256 =
+			keccak_256(pallet_assets::Pallet::<Runtime, Instance>::name(asset_id).as_slice())
```

Comment:
Names could be identical for different asset ids, maybe we should use the assetId itself? or is this something standard?

Reply:
Based on how we treat asset Ids, it is unlikely that accross different instances assetIds are identical (we Hash multilocation or a counter)

Reply:
I am actually assuming this is here so that one cannot do a replay attack across instances right? I am getting this wrong?

Reply:
The domain separator is also based on the address of the asset precompile, thus avoiding replays accross assets having the same name.

Reply:
but in that case, isnt the address enough? or is this part of a standard you are following?

Reply:
In that case, isnt the address enough? or is this part of a standard you are following?